File: CPP_benchmark/matrix_benchmark.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <chrono>
#include <random>

using namespace std;
using namespace chrono;

vector<vector<int>> multiply_matrices(const vector<vector<int>>& A, const vector<vector<int>>& B, int N) {
    vector<vector<int>> C(N, vector<int>(N, 0));
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            for (int k = 0; k < N; k++)
                C[i][j] += A[i][k] * B[k][j];
    return C;
}

int main(int argc, char* argv[]) {
    std::mt19937 rng(42);  // Fixed seed for reproducibility
    std::uniform_int_distribution<int> dist(0, 10000);
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <matrix_size>" << endl;
        return 1;
    }

    int N = stoi(argv[1]);
    vector<vector<int>> A(N, vector<int>(N, rand() % 10));
    vector<vector<int>> B(N, vector<int>(N, rand() % 10));

    auto start = high_resolution_clock::now();
    vector<vector<int>> C = multiply_matrices(A, B, N);
    auto end = high_resolution_clock::now();

    cout << "Matrix Multiplication | Size: " << N 
         << " | Time: " << duration<double, milli>(end - start).count() << " ms" << endl;
    return 0;
}

File: WASM/Java_Benchmark/java-wasm-app/src/main/java/com/example/MatrixBenchmark.java
package com.example;

import java.util.Random;

public class MatrixBenchmark {
    public static int[][] multiplyMatrices(int[][] A, int[][] B, int N) {
        int[][] C = new int[N][N];
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                for (int k = 0; k < N; k++)
                    C[i][j] += A[i][k] * B[k][j];
        return C;
    }

    public static void main(String[] args) {
        int N = Integer.parseInt(args[0]);
        Random rand = new Random(42);
        
        int[][] A = new int[N][N], B = new int[N][N];
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++) {
                A[i][j] = rand.nextInt(10000);
                B[i][j] = rand.nextInt(10000);
            }

        long start = System.nanoTime();
        int[][] C = multiplyMatrices(A, B, N);
        long end = System.nanoTime();

        System.out.printf("Matrix Multiplication | Size: %dx%d | Time: %.2f ms%n", N, N, (end - start) / 1e6);
    }
}
